### ISSUE TITLE
(pedantic) The `box_collection` docs are not quite correct.

### ISSUE THREAD

**Issue by Lokathor** [rust-lang/rust-clippy#9392]
### Description

Currently the docs for [box_collection](https://rust-lang.github.io/rust-clippy/master/index.html#box_collection) state:
```
## Why is this bad?
Collections already keeps their contents in a separate area on the heap. So if you Box them, you just add another level of indirection without any benefit whatsoever.
```

However, there is one clear benefit: Box<Collection> is one non-null pointer big instead of being 2-3 (or more!) pointers big. Particularly for error data, the extra layer of boxing to keep the error variant small outweighs the mild speed penalty.

**Comment by GeeWee** [comment]
Stumbled upon this as we faced the exact same thing with regards to errors, where we'd get the [large_enum_variant](https://rust-lang.github.io/rust-clippy/master/index.html#/large_enum_variant) error as our errors are bigger - we'd then box collections inside the error to get under the size threshold, and then get this new clippy error instead.

**Comment by Ten0** [comment]
Here's another way in which the docs about this lint are wrong: they explicitly state "anywhere in the code", but while the following code triggers the lint: 
```rust
struct X {
	values: Box<Vec<u32>>,
}
```
the following code doesn't:
```rust
fn f() {
	let _v: Box<Vec<u32>> = Box::new(Vec::new());
}
```

So it appears to [not actually be "anywhere in the code"](https://github.com/rust-lang/rust-clippy/blob/8769afffe6f434a63295195ec5c4dbcb1e0d0f0d/clippy_lints/src/types/mod.rs#L591).

NB: this would probably better be fixed by making it actually run on "anywhere in the code", by adding it [here](https://github.com/rust-lang/rust-clippy/blob/8769afffe6f434a63295195ec5c4dbcb1e0d0f0d/clippy_lints/src/types/mod.rs#L665).
